### PR TITLE
Split two ranges to read messages if offsetInFileOfOldestMessage > offsetInFileAtEndOfNewestMessage to avoid possible loop

### PR DESCRIPTION
**Pull request by jianjunwoo**
My concern is there is still this line of code: [`try reader.seek(to: FileHeader.expectedEndOfHeaderInFile)` ](https://github.com/dfed/CacheAdvance/pull/66/files#diff-90b1f5431bb05caf1445ee7ee42e2a66b2bfb2e6fad63f3266cace6cd1389fd6L68), which keeps the possibility to run into an infinite loop. 

Thus, I have a thought to fix this for `shouldOverwriteOldMessages` true | false without the seek back action in this func. 

1. When `shouldOverwriteOldMessages` is true and new_offset < old_offset , split the reader by two ranges according old message offset, new message offset. 
    1.1 Range 1 would be : from old message offset to file end.
    1.2 Range 2 would be: from top to newest offset
2. When `shouldOverwriteOldMessages` is false,  the reader read from top to end.
3. Both of the above situations, we never seek back for next read which can guarantee there is no loop.

@dfed  @bachand

### CODE REVIEW

File: Sources/CacheAdvance/CacheAdvance.swift [hunk 1]
```diff
@@ -172,10 +170,28 @@ public final class CacheAdvance<T: Codable> {
...
+                messages.append(try decoder.decode(T.self, from: encodedMessage))
+            }
+
+            // This is first range: | `expectedEndOfHeaderInFile` -> `offsetInFileAtEndOfNewestMessage`|
+            for encodedMessage in try reader.encodedMessagesFromOffset(
+                FileHeader.expectedEndOfHeaderInFile,
+                endOffset: reader.offsetInFileAtEndOfNewestMessage) {
+                messages.append(try decoder.decode(T.self, from: encodedMessage))
+            }
+        }
```
Comment (dfed) [inline Sources/CacheAdvance/CacheAdvance.swift:194]: This is a clever approach! I don't have time to dig to deeply into this PR this week, but I'll do my best to get to it next week. My initial impression is that this approach seems reasonable. At a minimum, I really like the code comment you've written here – they'll really help future maintainers (myself included!).

File: Sources/CacheAdvance/CacheAdvance.swift [hunk 2]
```diff
@@ -172,10 +170,28 @@ public final class CacheAdvance<T: Codable> {
...
+            for encodedMessage in try reader.encodedMessagesFromOffset(reader.offsetInFileOfOldestMessage,
+                                                                       endOffset: reader.offsetInFileAtEndOfNewestMessage) {
+                messages.append(try decoder.decode(T.self, from: encodedMessage))
+            }
+        } else {
+            // In this case, the messages could be split to two ranges
+            // | First Range | (GAP: ignore) | Second Range |
 
+            // This is second range: | `offsetInFileOfOldestMessage` -> EOF |
+            for encodedMessage in try reader.encodedMessagesFromOffset(reader.offsetInFileOfOldestMessage) {
```
Comment (dfed) [inline Sources/CacheAdvance/CacheAdvance.swift:184]: We know the `endOffset` here, right? Since we know the `maximumBytes` of the file?

File: Sources/CacheAdvance/CacheReader.swift [hunk 3]
```diff
@@ -44,29 +41,26 @@ final class CacheReader {
...
 
-    /// Returns the next encodable message, seeking to the beginning of the next message.
-    func nextEncodedMessage() throws -> Data? {
-        let startingOffset = offsetInFile
-
-        guard startingOffset != offsetInFileAtEndOfNewestMessage else {
-            // We're at the last message.
-            return nil
+    /// Returns the encodable messages in a range
+    func encodedMessagesFromOffset(_ startOffset: UInt64, endOffset: UInt64? = nil) throws -> [Data] {
```
Comment (dfed) [inline Sources/CacheAdvance/CacheReader.swift:45]: I think we can make `endOffset` required given my previous comment.

I'd also consider renaming `endOffset` here to be `maximumOffset` or something to make it more clear that it is the largest possible offset for the end of this message, but that the message may (or even likely!) end before that offset.

File: Sources/CacheAdvance/CacheReader.swift [hunk 4]
```diff
@@ -44,29 +41,26 @@ final class CacheReader {
...
-            let endOfMessage = startingOffset + UInt64(MessageSpan.storageLength) + UInt64(messageLength)
-            let startingOffsetIsBeforeEndOfNewestMessageAndDoesNotExceedEndOfNewestMessage = startingOffset < offsetInFileAtEndOfNewestMessage && endOfMessage <= offsetInFileAtEndOfNewestMessage
-            let startingOffsetIsAfterEndOfNewestMessageAndDoesNotExceedEndOfFile = offsetInFileAtEndOfNewestMessage < startingOffset && endOfMessage <= maximumBytes
-            guard
-                startingOffsetIsBeforeEndOfNewestMessageAndDoesNotExceedEndOfNewestMessage
-                    || startingOffsetIsAfterEndOfNewestMessageAndDoesNotExceedEndOfFile
-            else {
-                // The offsetInFileAtEndOfNewestMessage is incorrect. This likely occured due to a crash when writing our header file.
-                throw CacheAdvanceError.fileCorrupted
-            }
```
Comment (dfed) [inline Sources/CacheAdvance/CacheReader.swift:68]: can you help me understand why we no longer need these checks?

File: Sources/CacheAdvance/CacheReader.swift [hunk 5]
```diff
@@ -75,16 +69,7 @@ final class CacheReader {
...
-                // We started reading before the offset of the end of the newest message, therefore we expected a message to be read. We instead read an empty space, meaning that the file is corrupt.
-                throw CacheAdvanceError.fileCorrupted
-            }
-
-            // We know the next message is at the end of the file header. Let's seek to it.
-            try reader.seek(to: FileHeader.expectedEndOfHeaderInFile)
-
-            // We know there's a message to read now that we're at the start of the file.
-            return try nextEncodedMessage()
+            return nil
```
Comment (dfed) [inline Sources/CacheAdvance/CacheReader.swift:72]: can you add a code comment here to help me understand what case we're hitting here / why it is expected that we return `nil` here?

File: Tests/CacheAdvanceTests/CacheAdvanceTests.swift [hunk 6]
```diff
@@ -96,114 +96,31 @@ final class CacheAdvanceTests: XCTestCase {
         XCTAssertEqual(messages, [])
     }
 
-    func test_messages_whenOffsetInFileAtEndOfNewestMessageIsBeyondEndOfNewestMessageButBeforeEndOfFile_throwsFileCorrupted() throws {
```
Comment (dfed) [inline Tests/CacheAdvanceTests/CacheAdvanceTests.swift:99]: why is this test (and the deleted one below) no longer valid? I think testing these cases is still valuable, but I might be missing something.

File: Tests/CacheAdvanceTests/CacheAdvanceTests.swift [hunk 7]
```diff
@@ -96,114 +96,31 @@ final class CacheAdvanceTests: XCTestCase {
...
-        XCTAssertThrowsError(try corruptedReadingCache.messages()) {
-            XCTAssertEqual($0 as? CacheAdvanceError, CacheAdvanceError.fileCorrupted)
-        }
-    }
-
-    func test_messages_whenOffsetInFileAtEndOfNewestMessageIsBeforeEndOfNewestMessage_throwsFileCorrupted() throws {
-        let message: TestableMessage = "This is a test"
-        let maximumBytes = try requiredByteCount(for: [message])
+    func test_messages_throwsFileCorruptedWhenOffsetInFileAtEndOfNewsetMessageOutOfSync() throws {
+        let randomHighValue: UInt64 = 10_1000
```
Comment (dfed) [inline Tests/CacheAdvanceTests/CacheAdvanceTests.swift:100]: let's make this number less confusing (and fix an old mistake of mine) by doing either:
```suggestion
        let randomHighValue: UInt64 = 10_000
```
or:
```suggestion
        let randomHighValue: UInt64 = 101_000
```

**Review by dfed — COMMENTED** [review]
I like how you're thinking about this problem! I need to dig in more before I can approve, which I won't have time to do until next week, but I'm excited by this direction 😄. In the meantime, I've left some comments/questions.

File: CacheAdvance.podspec [hunk 8]
```diff
@@ -1,6 +1,6 @@
 Pod::Spec.new do |s|
```
Comment (dfed) [inline CacheAdvance.podspec:1]: Starting a thread for this discussion:

> My concern is there is still this line of code: [`try reader.seek(to: FileHeader.expectedEndOfHeaderInFile)` ](https://github.com/dfed/CacheAdvance/pull/66/files#diff-90b1f5431bb05caf1445ee7ee42e2a66b2bfb2e6fad63f3266cace6cd1389fd6L68), which keeps the possibility to run into an infinite loop. 

I don't think that line can lead to an infinite loop per the unit tests in #66. We will either read all messages and then hit the end and stop, or we will detect that the file is corrupted and stop.

That said, I think the approach you're taking here may be easier to maintain + understand than the approach I took in #66, so I'm open to this approach even if we're not solving an existing problem. However, I would feel much more confident with this approach if we didn't revert the tests introduced in #66.

File: Tests/CacheAdvanceTests/CacheAdvanceTests.swift (hunk 6 above)
Comment (jianjunwoo) [reply Tests/CacheAdvanceTests/CacheAdvanceTests.swift:99]: These tests should be kept. I reverted the last commit, this was deleted then.

File: CacheAdvance.podspec (hunk 8 above)
Comment (dfed) [reply CacheAdvance.podspec:1]: That said, this new code may have different behavior (maybe we don't throw `fileCorrupted` anymore?), so we may need to _update_ the tests from #66.

Comment (jianjunwoo) [reply CacheAdvance.podspec:1]: > However, I would feel much more confident with this approach if we didn't revert the tests introduced in https://github.com/dfed/CacheAdvance/pull/66.

That's helpful. Let me keep the tests.

Comment (dfed) [reply CacheAdvance.podspec:1]: Per https://github.com/dfed/CacheAdvance/pull/72#discussion_r1013594675, it sounds like we're intending to keep the tests from #66! Excited to see if they need to be modified to pass with this change 🙂

File: Sources/CacheAdvance/CacheAdvance.swift (hunk 1 above)
Comment (jianjunwoo) [reply Sources/CacheAdvance/CacheAdvance.swift:194]: > I like how you're thinking about this problem! I need to dig in more before I can approve, which I won't have time to do until next week, but I'm excited by this direction 😄. In the meantime, I've left some comments/questions.

This approach came in when I read the discussion in PR#64 again. really happy to see this move forward if you feel it right too : D

File: Sources/CacheAdvance/CacheAdvance.swift (hunk 2 above)
Comment (jianjunwoo) [reply Sources/CacheAdvance/CacheAdvance.swift:184]: The `maximumBytes` may larger then the end offset I think.

File: Sources/CacheAdvance/CacheReader.swift (hunk 4 above)
Comment (jianjunwoo) [reply Sources/CacheAdvance/CacheReader.swift:68]: I would like to say we are safe if do not seek back in this method.

File: Sources/CacheAdvance/CacheReader.swift (hunk 5 above)
Comment (jianjunwoo) [reply Sources/CacheAdvance/CacheReader.swift:73]: here we return nil to stop reading

File: Sources/CacheAdvance/CacheReader.swift [hunk 9]
```diff
@@ -44,29 +41,29 @@ final class CacheReader {
...
+    func encodedMessagesFromOffset(_ startOffset: UInt64, endOffset: UInt64? = nil) throws -> [Data] {
+        var encodedMessages = [Data]()
+        try reader.seek(to: startOffset)
+        while let data = try nextEncodedMessage() {
+            encodedMessages.append(data)
+            if let endOffset, offsetInFile >= endOffset {
+                break
+            }
+        }
+        if let endOffset, offsetInFile != endOffset {
```
Comment (jianjunwoo) [inline Sources/CacheAdvance/CacheReader.swift:57]: If use `maximumOffset` set default value for `endOffset` in the func's parameter, this line need change to `offsetInFile < endOffset`.  
If changed to this, we need to change some test cases, like `test_messages_whenOffsetInFileAtEndOfNewestMessageIsBeyondEndOfFile_throwsFileCorrupted`. That case won't get error when `offsetInFileAtEndOfNewestMessage` is beyond EOF.

File: Sources/CacheAdvance/CacheReader.swift (hunk 3 above)
Comment (jianjunwoo) [reply Sources/CacheAdvance/CacheReader.swift:48]: I comment about this at line: https://github.com/dfed/CacheAdvance/pull/72/files#r1013767862

File: Sources/CacheAdvance/CacheAdvance.swift (hunk 2 above)
Comment (dfed) [reply Sources/CacheAdvance/CacheAdvance.swift:184]: You are correct. The max bytes number is the furthest possible offset but messages within the file might (and in fact likely! end before that point.

File: Sources/CacheAdvance/CacheReader.swift (hunk 4 above)
Comment (dfed) [reply Sources/CacheAdvance/CacheReader.swift:68]: Got it. Because these guards are preventing the seeking behavior which we are no longer doing we don't need them.  🤔. This makes sense. Will dig in more next week!

File: Sources/CacheAdvance/CacheReader.swift (hunk 9 above)
Comment (dfed) [inline Sources/CacheAdvance/CacheReader.swift:57]: FYI we support all the way back to Xcode 11 (we do this to run tests on older iOS/macOS/etc versions), so we can't use fancy Xcode 14 Swift syntax just yet!

File: Sources/CacheAdvance/CacheReader.swift (hunk 4 above)
Comment (jianjunwoo) [reply Sources/CacheAdvance/CacheReader.swift:68]: Wish you have a nice weekend! and continue the discussion next week : )

**Comment by bachand** [comment]
I will dig into this next week as well. Thanks @jianjunwoo !

File: Sources/CacheAdvance/CacheAdvance.swift [hunk 10]
```diff
@@ -172,10 +170,28 @@ public final class CacheAdvance<T: Codable> {
         try header.checkFile()
 
         var messages = [T]()
-        while let encodedMessage = try reader.nextEncodedMessage() {
-            messages.append(try decoder.decode(T.self, from: encodedMessage))
-        }
+        // There is only one range: | `offsetInFileOfOldestMessage` -> `offsetInFileAtEndOfNewestMessage`|
+        if reader.offsetInFileOfOldestMessage < reader.offsetInFileAtEndOfNewestMessage {
```
Comment (bachand) [inline Sources/CacheAdvance/CacheAdvance.swift:174]: When the cache is empty I assume that `reader.offsetInFileOfOldestMessage == reader.offsetInFileAtEndOfNewestMessage`? I wonder if it would be best to explicitly handle the case of these two values being equal, to make this code easier to reason about.

Comment (bachand) [reply Sources/CacheAdvance/CacheAdvance.swift:174]: It would also be nice to validate my assumption that `reader.offsetInFileOfOldestMessage == reader.offsetInFileAtEndOfNewestMessage` when the cache is empty.

File: Sources/CacheAdvance/CacheAdvance.swift [hunk 11]
```diff
@@ -172,10 +170,28 @@ public final class CacheAdvance<T: Codable> {
...
+            for encodedMessage in try reader.encodedMessagesFromOffset(reader.offsetInFileOfOldestMessage,
+                                                                       endOffset: reader.offsetInFileAtEndOfNewestMessage) {
+                messages.append(try decoder.decode(T.self, from: encodedMessage))
+            }
+        } else {
+            // In this case, the messages could be split to two ranges
+            // | First Range | (GAP: ignore) | Second Range |
```
Comment (bachand) [inline Sources/CacheAdvance/CacheAdvance.swift:181]: Does there _need_ to be a gap? I think technically the first range could be contiguous with the second range. If it's true that the ranges can be contiguous we should indicate that the gap is optional, for extra clarity.

File: Sources/CacheAdvance/CacheReader.swift [hunk 12]
```diff
@@ -23,12 +23,9 @@ final class CacheReader {
 
     /// Creates a new instance of the receiver.
     ///
-    /// - Parameters:
-    ///   - file: The file URL indicating the desired location of the on-disk store. This file should already exist.
-    ///   - maximumBytes: The maximum size of the cache, in bytes.
```
Comment (bachand) [inline Sources/CacheAdvance/CacheReader.swift:28]: 👌

File: Sources/CacheAdvance/CacheReader.swift [hunk 13]
```diff
@@ -44,29 +41,29 @@ final class CacheReader {
...
+            }
+        }
+        if let endOffset = endOffset, offsetInFile != endOffset {
+            throw CacheAdvanceError.fileCorrupted
         }
+        return encodedMessages
+    }
 
+    /// Returns the next encodable message, seeking to the beginning of the next message.
+    func nextEncodedMessage() throws -> Data? {
```
Comment (bachand) [inline Sources/CacheAdvance/CacheReader.swift:64]: Can we make this private now?

File: Sources/CacheAdvance/CacheReader.swift [hunk 14]
```diff
@@ -44,29 +41,29 @@ final class CacheReader {
...
+    /// Returns the encodable messages in a range
+    ///
+    /// - Parameter startOffset: the start offset to start reading
+    /// - Parameter endOffset: the end offset to stop reading, the default value is EOF if the value is nil.
+    func encodedMessagesFromOffset(_ startOffset: UInt64, endOffset: UInt64? = nil) throws -> [Data] {
+        var encodedMessages = [Data]()
+        try reader.seek(to: startOffset)
+        while let data = try nextEncodedMessage() {
+            encodedMessages.append(data)
+            if let endOffset = endOffset, offsetInFile >= endOffset {
```
Comment (bachand) [inline Sources/CacheAdvance/CacheReader.swift:53]: I'm thinking about the case where `offsetInFile == endOffset` to make sure I agree that we want to break in that case.

When we write a message we update the offset in file at the end of the newest message to the offset of the writer after the write.

https://github.com/dfed/CacheAdvance/blob/54e61e242d51e7239bdcf9f4307cd8355b7e86ba/Sources/CacheAdvance/CacheAdvance.swift#L247-L252

I believe that the offset at the end of the newest message is the same as the offset of the start of the next range.

This seems to be confirmed by the [documentation](https://developer.apple.com/documentation/foundation/filehandle/1410936-write) of the write method.

Screen Shot 2022-11-07 at 2 49 04 PM

Comment (bachand) [reply Sources/CacheAdvance/CacheReader.swift:53]: @dfed is this the correct documentation to reference? I was surprised to see we are using a deprecated method, though I recall that there was some reason we were doing so.

File: Sources/CacheAdvance/CacheReader.swift [hunk 15]
```diff
@@ -75,16 +72,8 @@ final class CacheReader {
             return message
 
         case .emptyRead:
-            guard offsetInFileAtEndOfNewestMessage < startingOffset else {
-                // We started reading before the offset of the end of the newest message, therefore we expected a message to be read. We instead read an empty space, meaning that the file is corrupt.
-                throw CacheAdvanceError.fileCorrupted
-            }
-
-            // We know the next message is at the end of the file header. Let's seek to it.
-            try reader.seek(to: FileHeader.expectedEndOfHeaderInFile)
```
Comment (bachand) [inline Sources/CacheAdvance/CacheReader.swift:84]: I agree that the library is safer against unknown unknown errors without this line of code.

**Review by bachand — APPROVED** [review]
Great work @jianjunwoo !

My understanding of this PR is as follows... We aren't currently aware of any type of file corruption that could lead to an infinite loop. At the same time, the way that code is currently written means that an infinite loop remains possible. This PR refactors the library so that an infinite loop is not possible.

It's worth pointing out that so far we haven't (as far as I can recall) acknowledged the two different "modes" of CacheAdvance in the code. The code has been written to transparently handle both situations. There's something elegant with not explicitly acknowledging each case and writing code that will work in either. At the same time, I have found that I spend a lot of time each time I review changes to CacheReader.swift thinking about the different possibilities. I feel like this change makes it easier to consider the edge cases since they are more explicitly captured in code.

I really like that we didn't need to change any tests in this PR. That makes me confident in the change.

Let's give @dfed the time to review as well to make sure he's onboard before we move ahead 👍

File: Sources/CacheAdvance/CacheReader.swift (hunk 13 above)
Comment (jianjunwoo) [reply Sources/CacheAdvance/CacheReader.swift:64]: It could be private, let me do in next commit. 
I kept it here to make the diff more readable.

File: Sources/CacheAdvance/CacheReader.swift (hunk 14 above)
Comment (dfed) [reply Sources/CacheAdvance/CacheReader.swift:53]: these are the `write` methods we're using:
https://github.com/dfed/CacheAdvance/blob/3a1ed6f99c2a870dc09c3a8144a58f915757a569/Sources/CacheAdvance/FileHandleExtensions.swift#L33-L40

I should take a pass to update the `__` methods in this file and use the more modern ones. We are not using deprecated methods when newer ones are available.

File: Sources/CacheAdvance/CacheAdvance.swift (hunk 11 above)
Comment (dfed) [reply Sources/CacheAdvance/CacheAdvance.swift:181]: I'd need to dig in more, but we have a unit test where we documented that we expect there to be a gap and we do not believe the ranges can be contiguous:
https://github.com/dfed/CacheAdvance/blob/3a1ed6f99c2a870dc09c3a8144a58f915757a569/Tests/CacheAdvanceTests/CacheAdvanceTests.swift#L350-L389

File: Sources/CacheAdvance/CacheAdvance.swift (hunk 10 above)
Comment (dfed) [reply Sources/CacheAdvance/CacheAdvance.swift:174]: Per this comment block
https://github.com/dfed/CacheAdvance/blob/3a1ed6f99c2a870dc09c3a8144a58f915757a569/Tests/CacheAdvanceTests/CacheAdvanceTests.swift#L367-L368

when `reader.offsetInFileOfOldestMessage` == `reader.offsetInFileAtEndOfNewestMessage`, the file is empty. I like the idea of explicitly handling this case.

File: Sources/CacheAdvance/CacheAdvance.swift [hunk 16]
```diff
@@ -172,10 +170,28 @@ public final class CacheAdvance<T: Codable> {
...
 
+            // This is second range: | `offsetInFileOfOldestMessage` -> EOF |
+            for encodedMessage in try reader.encodedMessagesFromOffset(reader.offsetInFileOfOldestMessage) {
+                messages.append(try decoder.decode(T.self, from: encodedMessage))
+            }
+
+            // This is first range: | `expectedEndOfHeaderInFile` -> `offsetInFileAtEndOfNewestMessage`|
```
Comment (dfed) [inline Sources/CacheAdvance/CacheAdvance.swift:188]: nit!
```suggestion
            // This is first range: | `expectedEndOfHeaderInFile` -> `offsetInFileAtEndOfNewestMessage` |
```

File: Sources/CacheAdvance/CacheAdvance.swift (hunk 10 above)
Comment (dfed) [inline Sources/CacheAdvance/CacheAdvance.swift:174]: This is nitty, but, let's put the comment within the `if` statement:
```suggestion
        if reader.offsetInFileOfOldestMessage < reader.offsetInFileAtEndOfNewestMessage {
            // There is only one range: | `offsetInFileOfOldestMessage` -> `offsetInFileAtEndOfNewestMessage` |
```

File: Sources/CacheAdvance/CacheAdvance.swift [hunk 17]
```diff
@@ -172,10 +170,28 @@ public final class CacheAdvance<T: Codable> {
...
+            for encodedMessage in try reader.encodedMessagesFromOffset(
+                FileHeader.expectedEndOfHeaderInFile,
+                endOffset: reader.offsetInFileAtEndOfNewestMessage) {
```
Comment (dfed) [inline Sources/CacheAdvance/CacheAdvance.swift:191]: let's break this out into a few lines with a named variable:
```suggestion
            let newerMessages = try reader.encodedMessagesFromOffset(
                FileHeader.expectedEndOfHeaderInFile,
                endOffset: reader.offsetInFileAtEndOfNewestMessage)
            for encodedMessage in newerMessages {
```

File: Sources/CacheAdvance/CacheAdvance.swift (hunk 2 above)
Comment (dfed) [inline Sources/CacheAdvance/CacheAdvance.swift:184]: Let's break this out into a few lines with a named variable:
```suggestion
            let olderMessages = try reader.encodedMessagesFromOffset(reader.offsetInFileOfOldestMessage)
            for encodedMessage in olderMessages {
```

File: Sources/CacheAdvance/CacheAdvance.swift [hunk 18]
```diff
@@ -172,10 +170,28 @@ public final class CacheAdvance<T: Codable> {
         try header.checkFile()
 
         var messages = [T]()
-        while let encodedMessage = try reader.nextEncodedMessage() {
-            messages.append(try decoder.decode(T.self, from: encodedMessage))
-        }
+        // There is only one range: | `offsetInFileOfOldestMessage` -> `offsetInFileAtEndOfNewestMessage`|
+        if reader.offsetInFileOfOldestMessage < reader.offsetInFileAtEndOfNewestMessage {
+            for encodedMessage in try reader.encodedMessagesFromOffset(reader.offsetInFileOfOldestMessage,
+                                                                       endOffset: reader.offsetInFileAtEndOfNewestMessage) {
```
Comment (dfed) [inline Sources/CacheAdvance/CacheAdvance.swift:176]: Let's break this out into a few lines with a named variable:

```suggestion
            let encodedMessages = try reader.encodedMessagesFromOffset(
                reader.offsetInFileOfOldestMessage,
                endOffset: reader.offsetInFileAtEndOfNewestMessage)
            for encodedMessage in encodedMessages {
```

File: Sources/CacheAdvance/CacheReader.swift [hunk 19]
```diff
@@ -44,29 +41,29 @@ final class CacheReader {
...
-    /// Returns the next encodable message, seeking to the beginning of the next message.
-    func nextEncodedMessage() throws -> Data? {
-        let startingOffset = offsetInFile
-
-        guard startingOffset != offsetInFileAtEndOfNewestMessage else {
-            // We're at the last message.
-            return nil
+    /// Returns the encodable messages in a range
+    ///
+    /// - Parameter startOffset: the start offset to start reading
```
Comment (dfed) [inline Sources/CacheAdvance/CacheReader.swift:46]: nitty suggestion:
```suggestion
    /// - Parameter startOffset: the offset from which to start reading
```

File: Sources/CacheAdvance/CacheReader.swift [hunk 20]
```diff
@@ -44,29 +41,29 @@ final class CacheReader {
...
+    /// - Parameter endOffset: the end offset to stop reading, the default value is EOF if the value is nil.
```
Comment (dfed) [inline Sources/CacheAdvance/CacheReader.swift:47]: another nitty suggestion:
```suggestion
    /// - Parameter endOffset: the offset at which to stop reading. If `nil`, the end offset will be the EOF
```

File: Sources/CacheAdvance/CacheReader.swift (hunk 14 above)
Comment (dfed) [inline Sources/CacheAdvance/CacheReader.swift:53]: When `offsetInFile > endOffset`, wouldn't that mean that the file is corrupted?

EDIT: yup! you covered this case a few lines down 😄

File: Sources/CacheAdvance/CacheReader.swift (hunk 13 above)
Comment (dfed) [inline Sources/CacheAdvance/CacheReader.swift:64]: I really like that this method is no longer very smart. Our code is more declarative and less imperative now, and the new approach should be simpler to maintain 😄

File: Sources/CacheAdvance/CacheReader.swift [hunk 21]
```diff
@@ -75,16 +72,8 @@ final class CacheReader {
...
-                // We started reading before the offset of the end of the newest message, therefore we expected a message to be read. We instead read an empty space, meaning that the file is corrupt.
-                throw CacheAdvanceError.fileCorrupted
-            }
-
-            // We know the next message is at the end of the file header. Let's seek to it.
-            try reader.seek(to: FileHeader.expectedEndOfHeaderInFile)
-
-            // We know there's a message to read now that we're at the start of the file.
-            return try nextEncodedMessage()
+            // We hit an empty read when reach EOF, so return nil to stop.
```
Comment (dfed) [inline Sources/CacheAdvance/CacheReader.swift:75]: Thoughts on the following comment?
```suggestion
            // An empty read means we hit the EOF. It is the responsibility of the calling code to validate this assumption.
```

**Review by dfed — APPROVED** [review]
I left a ton of little nit suggestions, but overall I really dig this approach. I appreciate your spending the time to make this library easier to maintain!

I have also validated that the lines not covered by tests in `CacheReader.swift` are already not covered on `main`.

File: Sources/CacheAdvance/CacheReader.swift (hunk 14 above)
Comment (dfed) [reply Sources/CacheAdvance/CacheReader.swift:53]: Took a pass at updating the methods we use here: https://github.com/dfed/CacheAdvance/pull/74

File: Sources/CacheAdvance/CacheAdvance.swift (hunk 10 above)
Comment (jianjunwoo) [reply Sources/CacheAdvance/CacheAdvance.swift:174]: like this idea too. handle `==` as empty will make it more clear

File: Sources/CacheAdvance/CacheReader.swift [hunk 22]
```diff
@@ -114,6 +83,26 @@ final class CacheReader {
 
     // MARK: Private
 
+    /// Returns the next encodable message, seeking to the beginning of the next message.
+    private func nextEncodedMessage() throws -> Data? {
```
Comment (jianjunwoo) [inline Sources/CacheAdvance/CacheReader.swift:87]: Mark this func as private since it is not used by other files any more.

File: Sources/CacheAdvance/CacheAdvance.swift (hunk 11 above)
Comment (jianjunwoo) [reply Sources/CacheAdvance/CacheAdvance.swift:186]: The gap is generated when the new message offset override the old message. Usually the message length of new is not the same as the old message to be rewrote. Then there probably is a gap.

File: Sources/CacheAdvance/CacheReader.swift [hunk 23]
```diff
@@ -50,8 +50,12 @@ final class CacheReader {
...
-                break
+            if let endOffset = endOffset {
+                if offsetInFile == endOffset {
+                    break
+                } else if offsetInFile > endOffset {
+                    throw CacheAdvanceError.fileCorrupted
+                }
             }
         }
         if let endOffset = endOffset, offsetInFile != endOffset {
```
Comment (dfed) [inline Sources/CacheAdvance/CacheReader.swift:61]: do we need this `if` anymore given the `if` within the `while` loop?

Comment (jianjunwoo) [reply Sources/CacheAdvance/CacheReader.swift:61]: This make no difference compared to the code before this. However, this looks more readable to handle `==` and `>` separately. @dfed 

```
                else if offsetInFile > endOffset {
                    throw CacheAdvanceError.fileCorrupted
                }
```

Comment (jianjunwoo) [reply Sources/CacheAdvance/CacheReader.swift:61]: Oooh, I misunderstood your question. The `if` within the `while` loop is need to check `offsetInFile == endOffset `  to stop at the end.

File: Sources/CacheAdvance/CacheAdvance.swift (hunk 11 above)
Comment (bachand) [reply Sources/CacheAdvance/CacheAdvance.swift:186]: `test_append_inAnOverwritingCacheEvictsMinimumPossibleNumberOfMessages()` proves that there must be a gap. I'm resolving this comment thread.

File: Sources/CacheAdvance/CacheAdvance.swift [hunk 24]
```diff
@@ -172,10 +170,35 @@ public final class CacheAdvance<T: Codable> {
...
+            // There is only one range: | `offsetInFileOfOldestMessage` -> `offsetInFileAtEndOfNewestMessage` |
+            let encodedMessages = try reader.encodedMessagesFromOffset(
+                reader.offsetInFileOfOldestMessage,
+                endOffset: reader.offsetInFileAtEndOfNewestMessage)
+            for encodedMessage in encodedMessages {
+                messages.append(try decoder.decode(T.self, from: encodedMessage))
+            }
+        } else if reader.offsetInFileOfOldestMessage == reader.offsetInFileAtEndOfNewestMessage {
+            // This is an empty cache.
+            return []
```
Comment (bachand) [inline Sources/CacheAdvance/CacheAdvance.swift:183]: 👍

File: Sources/CacheAdvance/CacheReader.swift (hunk 23 above)
Comment (bachand) [reply Sources/CacheAdvance/CacheReader.swift:61]: > Ahh. This code is catching the `offsetInFile < endOffset` case. Maybe we could make that explicit here? Far from necessary, but it could make the intent more clear if I'm reading this right.

I agree it would be nice to make this more clear.

**Review by bachand — APPROVED** [review]
Nice!

**Comment by jianjunwoo** [comment]
Looks like we've all set about the details, or do we still have any other comments that I missed to discuss?

File: Sources/CacheAdvance/CacheReader.swift (hunk 23 above)
Comment (dfed) [reply Sources/CacheAdvance/CacheReader.swift:61]: I'll take this in a follow-up PR

**Comment by dfed** [comment]
That's it! Thanks again @jianjunwoo for tackling this + making future maintenance of this lib easier! Your contributions have immensely improved this library in recent weeks :pray:
